frontend/packages/sdk/src/OnChainRegistry.ts

```diff
@@ -0,0 +1,110 @@
+import type { ApiPromise } from '@polkadot/api';
+
+import { waitReady } from "@polkadot/wasm-crypto";
+import { BN } from '@polkadot/util';
+
+import { createPruntimeApi } from './create';
+import { pruntime_rpc } from './proto';
+
+export class UnexpectedEndpointError extends Error {
+}
+
+
+export class OnChainRegistry {
+
+  public api: ApiPromise
+
+  public clusterId: string | undefined
+  public clusterInfo: Record<string, any> | undefined
+  public remotePubkey: string | undefined
+  public pruntimeURL: string | undefined
+
+  #ready: boolean = false
+  #phactory: pruntime_rpc.PhactoryAPI | undefined
+
+  constructor(api: ApiPromise) {
+    this.api = api
+  }
+
+  public async waitReady() {
+    await waitReady();
+
+    const clusters: [string, any][] = []
+    {
+      const result = await this.api.query.phalaFatContracts.clusters.entries()
+      // @ts-ignore
+      result.forEach(([storageKey, value]) => {
+        const clusterId = storageKey.toHuman()
+      // @ts-ignore
+        const clusterInfo = value.unwrap().toJSON()
```

Comment:
Why `toJSON()`? If not, you can get the `gasPrice` BN object directly, instead of converting back and forth.